### PR TITLE
Added ability to set excluded incoming message headers from copying t…

### PR DESCRIPTION
**Pull request by bullmastiffo**
…o outgoing messages

### CODE REVIEW

File: Sources/Contour/Configuration/BusConfiguration.cs [hunk 1]
```diff
@@ -1,4 +1,6 @@
-﻿namespace Contour.Configuration
+﻿using Contour.Transport.RabbitMQ;
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/BusConfiguration.cs]: There should be no transport specific references here.

File: Sources/Contour/Configuration/BusConfiguration.cs [hunk 2]
```diff
@@ -158,6 +160,9 @@ public IEnumerable<ISenderConfiguration> SenderConfigurations
         /// </summary>
         public IPayloadConverter Serializer { get; private set; }
 
+        /// <inheritdoc />
+        public IEnumerable<string> ExcludedIncomingHeaders { get; private set; }
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/BusConfiguration.cs]: What is the configuration scope of this setting?
If it is Endpoint then it needs to be moved to EndpointOptions etc.

Comment (semion-akimtsev) [inline Sources/Contour/Configuration/BusConfiguration.cs]: I suggest that it should be renamed to ExcludeHeaders like "use", "reuseConnection" etc.

File: Sources/Contour/Configuration/IBusConfiguration.cs [hunk 3]
```diff
@@ -1,37 +1,37 @@
...
+
+        /// <summary>
+        ///   Сериализатор сообщений.
+        /// </summary>
+        IPayloadConverter Serializer { get; }
+
+        /// <summary>
+        /// Incoming headers that are excluded from copying to outgoing message
+        /// </summary>
+        IEnumerable<string> ExcludedIncomingHeaders { get; }
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/IBusConfiguration.cs]: This should be moved to EndpointOptions, SendersDefaults or ReceiverDefaults.

File: Sources/Contour/Configuration/IBusConfigurator.cs [hunk 4]
```diff
@@ -294,5 +296,12 @@ public interface IBusConfigurator
         /// <param name="endpointBuilder">
         /// </param>
         void UseSubscriptionEndpointBuilder(Func<ISubscriptionEndpointBuilder, ISubscriptionEndpoint> endpointBuilder);
+
+
+        /// <summary>
+        /// Sets excluded incoming message headers
+        /// </summary>
+        /// <param name="excludedHeaders">Excluded headers</param>
+        void SetExcludedIncomingHeaders(IEnumerable<string> excludedHeaders);
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/IBusConfigurator.cs:305]: This change should be made according to the scope of the ExcludeHeaders property.

File: Sources/Contour/Configurator/EndpointElement.cs [hunk 5]
```diff
@@ -175,5 +177,21 @@ public ValidatorCollection Validators
             }
         }
 
+        [ConfigurationProperty("excluded-headers")]
+        public string ExcludedHeadersString => (string)base["excluded-headers"];
+
+        public IEnumerable<string> ExcludedHeaders
```
Comment (semion-akimtsev) [inline Sources/Contour/Configurator/EndpointElement.cs]: This type is not responsible for string operations. It should only read the configuration. This should be moved to a place where a list of headers is obtained from the configuration property.

File: Sources/Contour/Transport/RabbitMQ/BusConfigurationEx.cs [hunk 6]
```diff
@@ -126,5 +113,31 @@ public static IBusConfigurator UseRabbitMq(this IBusConfigurator busConfigurator
...
+                                         Headers.Expires,
+                                         Headers.MessageLabel,
+                                         Headers.Persist,
+                                         Headers.QueueMessageTtl,
+                                         Headers.ReplyRoute,
+                                         Headers.Timeout,
+                                         Headers.Ttl
+                                     };
+            blockedHeaders.AddRange(excludedIncomingHeaders ?? Enumerable.Empty<string>());
+            var messageHeaderStorage = new Maybe<IIncomingMessageHeaderStorage>(new MessageHeaderStorage(blockedHeaders));
```
Comment (semion-akimtsev) [inline Sources/Contour/Transport/RabbitMQ/BusConfigurationEx.cs]: Maybe is obsolete.

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 7]
```diff
@@ -917,7 +1010,7 @@ public void should_be_default()
         /// </summary>
         [TestFixture]
         [Category("Unit")]
-        public class when_configuring_endpoint_with_parallelism_level
+        public class c_with_parallelism_level
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: ?

**Review by semion-akimtsev — COMMENTED** [review]
Please review the comments.

File: Sources/Contour/Configuration/BusConfiguration.cs (hunk 1 above)
Comment (bullmastiffo) [reply Sources/Contour/Configuration/BusConfiguration.cs]: moved

File: Sources/Contour/Configuration/BusConfiguration.cs (hunk 2 above)
Comment (bullmastiffo) [reply Sources/Contour/Configuration/BusConfiguration.cs]: removed setting

Comment (bullmastiffo) [reply Sources/Contour/Configuration/BusConfiguration.cs]: removed setting

File: Sources/Contour/Configuration/IBusConfiguration.cs (hunk 3 above)
Comment (bullmastiffo) [reply Sources/Contour/Configuration/IBusConfiguration.cs]: removed setting

File: Sources/Contour/Configuration/IBusConfigurator.cs (hunk 4 above)
Comment (bullmastiffo) [reply Sources/Contour/Configuration/IBusConfigurator.cs:305]: There's no scope for endpoint configuration, this improvement is not a subject to change endpoint configuration

File: Sources/Contour/Configurator/EndpointElement.cs (hunk 5 above)
Comment (bullmastiffo) [reply Sources/Contour/Configurator/EndpointElement.cs]: It's only reading properties. No extra operations. String operations are required to manage serialization used in this config section.

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 7 above)
Comment (bullmastiffo) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: oops, fixed

File: Sources/Contour/Configuration/BusConfiguration.cs [hunk 8]
```diff
@@ -555,6 +553,25 @@ public void UseRouteResolverBuilder(Func<IRouteResolverBuilder, IRouteResolver>
...
+                                         Headers.MessageLabel,
+                                         Headers.Persist,
+                                         Headers.QueueMessageTtl,
+                                         Headers.ReplyRoute,
+                                         Headers.Timeout,
+                                         Headers.Ttl
+                                     };
+            blockedHeaders.AddRange(excludedHeaders ?? Enumerable.Empty<string>());
+            var messageHeaderStorage = new Maybe<IIncomingMessageHeaderStorage>(new MessageHeaderStorage(blockedHeaders));
+            this.SenderDefaults.IncomingMessageHeaderStorage = messageHeaderStorage;
```
Comment (mzabolotko) [inline Sources/Contour/Configuration/BusConfiguration.cs]: I believe the SetExcludedIncomingHeaders method is not right place to set  MessageHeaderStorage. Maybe replace this by the existing method WithIncomingMessageHeaderStorage using overload version with extra parameter 'blockingHeaders'?

Comment (bullmastiffo) [reply Sources/Contour/Configuration/BusConfiguration.cs]: adding to storage through interface method

File: Sources/Contour/Configuration/IBusConfigurator.cs (hunk 4 above)
Comment (semion-akimtsev) [reply Sources/Contour/Configuration/IBusConfigurator.cs:305]: Endpoint, incoming and outgoing settings represent the corresponding configuration scopes.

File: Sources/Contour/Configurator/EndpointElement.cs (hunk 5 above)
Comment (semion-akimtsev) [reply Sources/Contour/Configurator/EndpointElement.cs]: Reading DOES NOT involve processing and splitting a string.

File: Sources/Contour/Configuration/BusConfiguration.cs [hunk 9]
```diff
@@ -560,12 +553,23 @@ public void UseRouteResolverBuilder(Func<IRouteResolverBuilder, IRouteResolver>
...
-        }
-
+        }
+
+        /// <inheritdoc />
         public void SetExcludedIncomingHeaders(IEnumerable<string> excludedHeaders)
         {
-            this.ExcludedIncomingHeaders = excludedHeaders;
-            this.UseMessageHeaderStorage(this.ExcludedIncomingHeaders);
+            var blockedHeaders = (excludedHeaders ?? Enumerable.Empty<string>()).ToArray();
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/BusConfiguration.cs]: What's the difference between blocked and excluded?

File: Sources/Contour/Configuration/BusConfiguration.cs [hunk 10]
```diff
@@ -560,12 +553,23 @@ public void UseRouteResolverBuilder(Func<IRouteResolverBuilder, IRouteResolver>
...
+            var strg = this.SenderDefaults.GetIncomingMessageHeaderStorage();
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/BusConfiguration.cs]: Are all the vowels gone?

File: Sources/Contour/IIncomingMessageHeaderStorage.cs [hunk 11]
```diff
@@ -18,5 +18,11 @@ public interface IIncomingMessageHeaderStorage
         /// </summary>
         /// <returns>Заголовки входящего сообщения.</returns>
         IDictionary<string, object> Load();
+
+        /// <summary>
+        /// Registers header names to be excluded on storing
+        /// </summary>
+        /// <param name="headers">Header names</param>
+        void RegisterExcludedHeaders(IEnumerable<string> headers);
```
Comment (semion-akimtsev) [inline Sources/Contour/IIncomingMessageHeaderStorage.cs:26]: The storage is supposed to store and load. If some headers need to be excluded they should be passed via constructor.

File: Sources/Contour/Configurator/EndpointElement.cs [hunk 12]
```diff
@@ -175,5 +177,21 @@ public ValidatorCollection Validators
             }
         }
 
+        [ConfigurationProperty("excluded-headers")]
```
Comment (semion-akimtsev) [inline Sources/Contour/Configurator/EndpointElement.cs]: excludedHeaders

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 13]
```diff
@@ -908,7 +980,31 @@ public void should_be_default()
                     Assert.IsTrue(qosMaybe.HasValue, "QoS должен быть установлен.");
                     Assert.AreEqual(50, qosMaybe.Value.PrefetchCount, "Должно быть установлено количество потоков.");
                 }
+            }
+
+            [Test]
+            public void should_set_excluded_headers()
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs:983]: What is the SUT here?

**Review by semion-akimtsev — CHANGES_REQUESTED** [review]
Please make the changes and resubmit.

File: Sources/Contour/Configurator/EndpointElement.cs (hunk 5 above)
Comment (bullmastiffo) [reply Sources/Contour/Configurator/EndpointElement.cs]: Reading settings DOES

File: Tests/Contour.Common.Tests/MessageHeaderStorageSpecs.cs [hunk 14]
```diff
@@ -22,14 +22,22 @@ public class WhenStoreHeaders
...
 
-                var sut = new MessageHeaderStorage(blackList);
+                var excludedListAddition = new List<string> { "e", "f" };
 
-                sut.Store(new Dictionary<string, object> { { "a", 1 }, { "b", 2 }, { "c", 3 }, { "d", 4 } });
+                var sut = new MessageHeaderStorage(excludedList);
 
-                CollectionAssert.DoesNotContain(sut.Load().Keys, "b", "Заголовок должен быть исключен при сохранении.");
-                CollectionAssert.DoesNotContain(sut.Load().Keys, "c", "Заголовок должен быть исключен при сохранении.");
+                sut.RegisterExcludedHeaders(excludedListAddition);
```
Comment (semion-akimtsev) [inline Tests/Contour.Common.Tests/MessageHeaderStorageSpecs.cs:31]: This should be validated separately.

**Review by semion-akimtsev — APPROVED** [review]

**Comment by semion-akimtsev** [comment]
@bullmastiffo @mzabolotko 

Hi guys,
just to notify you - I have made a minor change to the sender and receiver options to support explicit message storage configuration on the endpoint and its propagation to the Rabbit specific defaults. See the commit below https://github.com/SDVentures/Contour/commit/c7cadc07152afef8cde541fa10cbe9633b0ef6ef.

Please do not hesitate to post questions and concerns.

**Pull request merged**